Feature: Kubernetes endpointslices to scale-out

szuecs opened the pull request:
Feature: Kubernetes endpointslices to scale out services to more than 1000 pods 
Related risk: [#427](https://github.bus.zalan.do/risk-register/zalando/issues/427)

szuecs commented on dataclients/kubernetes/clusterclient.go:
```diff
@@ -486,9 +580,23 @@ func (c *clusterClient) fetchClusterState() (*clusterState, error) {
+	// if err != nil {
+	// 	return nil, err
+	// }
+	if endpoints == nil && endpointSlices == nil {
```
any idea how to solve this good?
Basically it's enough to fetch one of these (endpointslices or endpoints) if for example permissions are not set correctly or we run in a cluster that has no endpointslices.

AlexanderYastrebov replied:
Lets do endpoint-slices flag. if false fetch and use endpoints otherwise endpointslices

szuecs replied:
done

AlexanderYastrebov commented on dataclients/kubernetes/ingressdefinitions.go:
```diff
@@ -65,6 +65,145 @@ func (s service) getTargetPortByValue(p int) (*definitions.BackendPort, bool) {
+}
+
+type endpointSliceList struct {
+	Meta  *definitions.ListMetadata
```
Should it be tagged as "`json:",inline"`"?

Maybe we can just use existing metadata here and simply add continue and remaining fields without introducing ListMetadata type?

AlexanderYastrebov replied:
I see now that continue field is part of metadata. I think it should be tagged with "`json:"metadata"`"

szuecs replied:
We don't need "continue". It was a miss understanding by myself but I clarified with sig-network.
The "continue" only exists in Lists because it is about number of objects in lists (for example GET all pods and if this listing is more than 1000 objects). One endpointslice holds a list of endpoints but it not a list request but a GET of a single object.

szuecs replied:
this will addressed in the next push

AlexanderYastrebov commented on dataclients/kubernetes/clusterclient.go:
```diff
@@ -449,6 +457,92 @@ func (c *clusterClient) loadEndpoints() (map[definitions.ResourceID]*endpoint, e
+func (c *clusterClient) loadEndpointSlices() (map[definitions.ResourceID]*skipperEndpointSlice, error) {
+	var endpointSlices endpointSliceList
+
+	if err := c.getJSON(c.endpointSlicesURI+c.endpointSlicesLabelSelectors, &endpointSlices); err != nil {
```
I think this should be a loop while .metadata.continue field is not empty.
It also should set default fetch limit.

AlexanderYastrebov replied:
I think it also makes sense to fetch all pages of k8s objects first and then process/transform them in to return result in a separate loop.

szuecs replied:
https://github.com/zalando/skipper/issues/2574

AlexanderYastrebov commented on dataclients/kubernetes/definitions/definitions.go:
```diff
@@ -9,6 +9,16 @@ import (
 var errInvalidMetadata = errors.New("invalid metadata")
 
+type ListMetadata struct {
+	*Metadata
```
I think ListMetadata does not embed standard metadata

szuecs replied:
this will addressed in the next push

AlexanderYastrebov commented on dataclients/kubernetes/ingressdefinitions.go:
```diff
@@ -65,6 +65,145 @@ func (s service) getTargetPortByValue(p int) (*definitions.BackendPort, bool) {
+
+// see https://kubernetes.io/docs/reference/kubernetes-api/service-resources/endpoint-slice-v1/#EndpointSlice
+type endpointSlice struct {
+	Meta        *definitions.Metadata     `json:"Metadata"`
```
Lowercase "metadata"

szuecs replied:
this will addressed in the next push

AlexanderYastrebov commented on dataclients/kubernetes/ingressdefinitions.go:
```diff
@@ -65,6 +65,145 @@ func (s service) getTargetPortByValue(p int) (*definitions.BackendPort, bool) {
+// Kubernetes endpointslices of a given service can have duplicates with different states.
+// Therefore Kubernetes endpointslices need to be de-duplicated before usage.
+// The business object skipperEndpointSlice is a de-duplicated endpoint list that concats all endpointslices of a given service into one slice of skipperEndpointSlice.
+type skipperEndpointSlice struct {
```
Lets move this into separate enpointslices.go file

szuecs replied:
endpoints are also not in a separated file. I will split out both endpoints.go and endpointslices.go

szuecs replied:
done

AlexanderYastrebov commented on dataclients/kubernetes/ingressdefinitions.go:
```diff
@@ -65,6 +65,145 @@ func (s service) getTargetPortByValue(p int) (*definitions.BackendPort, bool) {
+	Zone    string
+}
+
+func (eps skipperEndpointSlice) getPort(protocol, pName string, pValue int) int {
```
Lets use pointer receiver following https://dave.cheney.net/2016/03/19/should-methods-be-declared-on-t-or-t

szuecs replied:
But we use value receiver in all the other functions, too (service, endpoints). Do you mean I should change all?
To have only one different from the other makes no sense IMO.

AlexanderYastrebov replied:
Up to you

szuecs replied:
done

AlexanderYastrebov commented on dataclients/kubernetes/kubernetestest/api.go:
```diff
@@ -281,6 +284,10 @@ func initNamespace(kinds map[string][]interface{}) (ns namespace, err error) {
 		return
 	}
 
+	if err = itemsJSON(&ns.endpointslices, kinds["EndpointSlice"]); err != nil {
```
Should we model pagination here?

szuecs replied:
It would be nice to have but it's not the problem we need to solve, see also the comment above

szuecs replied:
https://github.com/zalando/skipper/issues/2574

AlexanderYastrebov commented on dataclients/kubernetes/clusterclient.go:
```diff
@@ -449,6 +459,88 @@ func (c *clusterClient) loadEndpoints() (map[definitions.ResourceID]*endpoint, e
+				// if conditions are nil then we need to treat is as ready
+				if ep.Conditions == nil {
+					resEps[ep.Addresses[0]] = &skipperEndpoint{
+						// Addresses [1..100] of the same AddressType, as kube-proxy we use the first
+						// see also https://github.com/kubernetes/kubernetes/issues/106267
+						Address: ep.Addresses[0],
```
Lets intriduce local variable `address := ep.Addresses[0]` with the comment and use it everywhere

szuecs replied:
done

AlexanderYastrebov commented on dataclients/kubernetes/clusterclient.go:
```diff
@@ -449,6 +459,88 @@ func (c *clusterClient) loadEndpoints() (map[definitions.ResourceID]*endpoint, e
+
+				if ep.isTerminating() {
+					terminatingEps[ep.Addresses[0]] = struct{}{}
+					// if we had this one with a non termiating condition, we should delete it, because of eventual consistency it is actually terminating
```
termiating -> terminating

szuecs replied:
done

AlexanderYastrebov commented on dataclients/kubernetes/clusterclient.go:
```diff
@@ -449,6 +459,88 @@ func (c *clusterClient) loadEndpoints() (map[definitions.ResourceID]*endpoint, e
+						Address: ep.Addresses[0],
+						Zone:    ep.Zone,
+					}
+					continue
```
Lets use `else if` instead of continue

szuecs replied:
done

AlexanderYastrebov commented on dataclients/kubernetes/clusterclient.go:
```diff
@@ -449,6 +459,88 @@ func (c *clusterClient) loadEndpoints() (map[definitions.ResourceID]*endpoint, e
+					delete(resEps, ep.Addresses[0])
+					continue
+				}
+				// already known terminating?
+				if _, ok := terminatingEps[ep.Addresses[0]]; ok {
+					continue
+				}
```
Does this mean it was observed termination but now it is not (because previous ep.isTerminating() was false)?

AlexanderYastrebov replied:
Should we log a warning maybe?

szuecs replied:
We do not want to have endpoints with `Terminating: true` in the list of load balancer members.
I don't expect a terminating pods to become `Terminating: false` again.
Endpointslices can have duplicate entries with different states, so if you have 2 slices and one endpoint in both (think also eventual consistency), it should be enough to know that one is terminating to drop it out of the member list.

szuecs replied:
I had to change the order of the logic based on testing input.

AlexanderYastrebov commented on dataclients/kubernetes/clusterclient.go:
```diff
@@ -449,6 +459,88 @@ func (c *clusterClient) loadEndpoints() (map[definitions.ResourceID]*endpoint, e
+		}
+
+		for i := range epSlices {
+			resEps := make(map[string]*skipperEndpoint)
+			terminatingEps := make(map[string]struct{})
```
Is the scope of these correct? They will be reset on each new slice...

szuecs replied:
I guess no, I think there is a missing test coverage here, too!

szuecs replied:
fixed

AlexanderYastrebov commented on dataclients/kubernetes/endpointslices.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+func (eps *skipperEndpointSlice) targets(protocol, backendProtocol string) []string {
+	result := make([]string, 0)
```
We can pre-allocate capacity of len(eps.Endpoints) here

szuecs replied:
done

AlexanderYastrebov commented on dataclients/kubernetes/kube.go:
```diff
@@ -564,7 +573,7 @@ func (c *Client) GetEndpointAddresses(ns, name string) []string {
 	if c.state == nil {
 		return nil
 	}
-	return c.state.GetEndpointsByName(ns, name, "TCP")
```
This "TCP" looks suspicious.
We use GetEndpointAddresses only for redis updater and then trim "TCP://" prefix.
For endpoint slices now I see you need "TCP" too lookup port and "http" to format url.
Lets rename `backendProtocol` to `scheme` here and elsewhere.
I also think this change may break Redis updater because result will now have "http://" prefix, please check.

szuecs replied:
The redis updater call path was also changed to trim http://. I think I want to re-do this in the future. "TCP" comes from the Kubernetes service (you can basically have their the values "TCP" or "UDP" or "STCP" and in our case it's always "TCP", because we do not support http3 nor quic as client).
On the other hand the backendProtocol (or scheme) is used to add the scheme to endpoints which is in general useful (and we might be able to get this from "appProtocol" in endpointslices) but for simplicity and as we support only http I think it's ok to have it hard coded right now.

Will rename backendProtocol->scheme

szuecs replied:
done

AlexanderYastrebov commented on config/config.go:
```diff
@@ -787,6 +789,7 @@ func (c *Config) ToOptions() skipper.Options {
 		WhitelistedHealthCheckCIDR:             whitelistCIDRS,
 		KubernetesPathMode:                     c.KubernetesPathMode,
 		KubernetesNamespace:                    c.KubernetesNamespace,
+		KubernetesEnableEndpointslices:         c.KubernetesEnableEndpointSlices,
```
We also need to update KubernetesDataClientOptions() to have this flag in routesrv.

szuecs replied:
this was already done https://github.com/zalando/skipper/pull/2565/files#diff-9b0b16d06f27b640eb77f6ba0dee84df6a660528e5cccd81839e0330c4b92dd2R923

szuecs commented:
Rebased and squashed most of the commits. Not sure if the refactoring commits are better to split out, but I can also rebase it into a single commit for the release notes.
wdyt @AlexanderYastrebov what would fit better the release notes workflow?

ThorbjoernG commented:
It's a pretty big change that touches some key features. I would be in favor of breaking it up into smaller changes and rolling out iterative.
changing the VERSION sounds reasonable.

szuecs commented:
Splitting is not really possible. Test coverage is ~90% and deployment will be validated before rollout.
See also https://github.com/zalando-incubator/kubernetes-on-aws/pull/6434 which will deploy the new feature as shadow stack and we can validate that we get the same results before rolling out to production.

szuecs commented on dataclients/kubernetes/endpoints.go:
```diff
@@ -0,0 +1,111 @@
+}
+
+type endpoint struct {
+	Meta    *definitions.Metadata `json:"Metadata"`
```
Great spot! Oops, we had it since some time: https://github.com/zalando/skipper/blob/master/dataclients/kubernetes/ingressdefinitions.go#L41

fixed

szuecs commented:
CodeQL says: 
```
RequestError [HttpError]: API rate limit exceeded for installation ID 5260298. If you reach out to GitHub Support for help, please include the request ID 6463:243E:62D4D2:663DDE:65005276.
```

AlexanderYastrebov commented:
Rebased on top of master

szuecs commented:
Test runs comparing 200 clusters showed that the change seems to be creating the same routes.

szuecs commented:
:+1:

szuecs commented:
:+1:

AlexanderYastrebov commented:
:+1:

szuecs commented:
:+1:

AlexanderYastrebov commented:
:+1:
